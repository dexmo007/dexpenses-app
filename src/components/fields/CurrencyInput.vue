<template>
  <v-select
    :value="value"
    @input="$emit('input', $event)"
    :items="currencies"
    menu-props="auto"
    label="Currency"
    hide-details
    single-line
  ></v-select>
</template>
<script>
import { currencies } from '@/util/receipt';

export default {
  name: 'CurrencyInput',
  props: ['value'],
  data() {
    return {
      currencies,
    };
  },
};
</script>
