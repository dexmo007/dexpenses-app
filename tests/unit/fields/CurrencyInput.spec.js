import { mount, createLocalVue } from '@vue/test-utils';
import createVuetify from '../vuetify';
import CurrencyInput from '@/components/fields/CurrencyInput.vue';

const localVue = createLocalVue();

describe('CurrencyInput.vue', () => {
  let vuetify;
  beforeEach(() => {
    vuetify = createVuetify();
  });

  it('should render correctly', async () => {
    const wrapper = mount(CurrencyInput, {
      propsData: {
        eager: true, // so the selection option render without clicking the select
      },
      localVue,
      vuetify,
      attachToDocument: true,
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should render correctly with value', () => {
    const wrapper = mount(CurrencyInput, {
      propsData: {
        value: 'EUR',
        eager: true,
      },
      localVue,
      vuetify,
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should trigger input event if user selects new item', () => {
    const wrapper = mount(CurrencyInput, { localVue, vuetify });
    wrapper.find('.v-select__slot').trigger('click');
    wrapper.find('.v-list-item').trigger('click'); // click first item: EUR
    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.emitted().input).toBeTruthy();
    expect(wrapper.emitted().input[0]).toHaveLength(1);
    expect(wrapper.emitted().input[0][0]).toEqual('EUR');
  });

  it('should not trigger input event if new value is set programmatically', () => {
    const wrapper = mount(CurrencyInput, {
      propsData: {
        value: 'EUR',
        eager: true,
      },
      localVue,
      vuetify,
    });
    expect(wrapper.element).toMatchSnapshot();
    wrapper.setProps({
      value: 'USD',
    });
    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.emitted().input).toBeFalsy();
  });

  it('should update v-model correctly', () => {
    const wrapper = mount(
      {
        template: '<CurrencyInput v-model="modelValue" eager></CurrencyInput>',
        data: () => ({
          modelValue: null,
        }),
        components: {
          CurrencyInput,
        },
      },
      { localVue, vuetify }
    );

    wrapper.find('.v-select__slot').trigger('click');
    wrapper.find('.v-list-item').trigger('click'); // click first item: EUR

    expect(wrapper.vm.modelValue).toEqual('EUR');
    expect(wrapper.element).toMatchSnapshot();
  });
});
