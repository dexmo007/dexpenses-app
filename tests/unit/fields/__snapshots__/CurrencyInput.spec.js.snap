// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyInput.vue should not trigger input event if new value is set programmatically 1`] = `
<div
  class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--single-line v-select"
  eager="true"
>
  <div
    class="v-input__control"
  >
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-owns="list-42"
      class="v-input__slot"
      role="button"
    >
      <div
        class="v-select__slot"
      >
        <div
          class="v-select__selections"
        >
          <div
            class="v-select__selection v-select__selection--comma"
          >
            EUR
          </div>
          <input
            aria-readonly="true"
            id="input-42"
            readonly="readonly"
            type="text"
          />
        </div>
        <div
          class="v-input__append-inner"
        >
          <div
            class="v-input__icon v-input__icon--append"
          >
            <i
              aria-hidden="true"
              class="v-icon notranslate material-icons theme--light"
            >
              arrow_drop_down
            </i>
          </div>
        </div>
      </div>
      <div
        class="v-menu"
      >
        <div
          class="v-menu__content theme--light v-menu__content--auto "
          style="max-height: 200px; min-width: 0px; max-width: auto; top: 12px; left: 12px; transform-origin: top left; z-index: 0; display: none;"
        >
          <div
            class="v-select-list v-card theme--light"
            id="list-42"
          >
            <div
              class="v-list v-sheet v-sheet--tile theme--light"
              id="list-42"
              role="listbox"
              tabindex="-1"
            >
              <div
                aria-labelledby="eur-list-item-47"
                aria-selected="true"
                class="v-list-item primary--text v-list-item--active v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                <div
                  class="v-list-item__content"
                >
                  <div
                    class="v-list-item__title"
                    id="eur-list-item-47"
                  >
                    EUR
                  </div>
                </div>
              </div>
              <div
                aria-labelledby="usd-list-item-47"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                <div
                  class="v-list-item__content"
                >
                  <div
                    class="v-list-item__title"
                    id="usd-list-item-47"
                  >
                    USD
                  </div>
                </div>
              </div>
              <div
                aria-labelledby="gbp-list-item-47"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                <div
                  class="v-list-item__content"
                >
                  <div
                    class="v-list-item__title"
                    id="gbp-list-item-47"
                  >
                    GBP
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`CurrencyInput.vue should not trigger input event if new value is set programmatically 2`] = `
<div
  class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--single-line v-select"
  eager="true"
>
  <div
    class="v-input__control"
  >
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-owns="list-42"
      class="v-input__slot"
      role="button"
    >
      <div
        class="v-select__slot"
      >
        <div
          class="v-select__selections"
        >
          <div
            class="v-select__selection v-select__selection--comma"
          >
            USD
          </div>
          <input
            aria-readonly="true"
            id="input-42"
            readonly="readonly"
            type="text"
          />
        </div>
        <div
          class="v-input__append-inner"
        >
          <div
            class="v-input__icon v-input__icon--append"
          >
            <i
              aria-hidden="true"
              class="v-icon notranslate material-icons theme--light"
            >
              arrow_drop_down
            </i>
          </div>
        </div>
      </div>
      <div
        class="v-menu"
      >
        <div
          class="v-menu__content theme--light v-menu__content--auto "
          style="max-height: 200px; min-width: 0px; max-width: auto; top: 12px; left: 12px; transform-origin: top left; z-index: 0; display: none;"
        >
          <div
            class="v-select-list v-card theme--light"
            id="list-42"
          >
            <div
              class="v-list v-sheet v-sheet--tile theme--light"
              id="list-42"
              role="listbox"
              tabindex="-1"
            >
              <div
                aria-labelledby="eur-list-item-47"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                <div
                  class="v-list-item__content"
                >
                  <div
                    class="v-list-item__title"
                    id="eur-list-item-47"
                  >
                    EUR
                  </div>
                </div>
              </div>
              <div
                aria-labelledby="usd-list-item-47"
                aria-selected="true"
                class="v-list-item primary--text v-list-item--active v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                <div
                  class="v-list-item__content"
                >
                  <div
                    class="v-list-item__title"
                    id="usd-list-item-47"
                  >
                    USD
                  </div>
                </div>
              </div>
              <div
                aria-labelledby="gbp-list-item-47"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                <div
                  class="v-list-item__content"
                >
                  <div
                    class="v-list-item__title"
                    id="gbp-list-item-47"
                  >
                    GBP
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`CurrencyInput.vue should render correctly 1`] = `
<div
  class="v-input v-input--hide-details theme--light v-text-field v-text-field--single-line v-select"
  eager="true"
>
  <div
    class="v-input__control"
  >
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-owns="list-3"
      class="v-input__slot"
      role="button"
    >
      <div
        class="v-select__slot"
      >
        <label
          class="v-label theme--light"
          for="input-3"
          style="left: 0px; position: absolute;"
        >
          Currency
        </label>
        <div
          class="v-select__selections"
        >
          <input
            aria-readonly="true"
            id="input-3"
            readonly="readonly"
            type="text"
          />
        </div>
        <div
          class="v-input__append-inner"
        >
          <div
            class="v-input__icon v-input__icon--append"
          >
            <i
              aria-hidden="true"
              class="v-icon notranslate material-icons theme--light"
            >
              arrow_drop_down
            </i>
          </div>
        </div>
      </div>
      <div
        class="v-menu"
      >
        <div
          class="v-menu__content theme--light v-menu__content--auto "
          style="max-height: 200px; min-width: 0px; max-width: auto; top: 12px; left: 12px; transform-origin: top left; z-index: 0; display: none;"
        >
          <div
            class="v-select-list v-card theme--light"
            id="list-3"
          >
            <div
              class="v-list v-sheet v-sheet--tile theme--light"
              id="list-3"
              role="listbox"
              tabindex="-1"
            >
              <div
                aria-labelledby="eur-list-item-8"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                <div
                  class="v-list-item__content"
                >
                  <div
                    class="v-list-item__title"
                    id="eur-list-item-8"
                  >
                    EUR
                  </div>
                </div>
              </div>
              <div
                aria-labelledby="usd-list-item-8"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                <div
                  class="v-list-item__content"
                >
                  <div
                    class="v-list-item__title"
                    id="usd-list-item-8"
                  >
                    USD
                  </div>
                </div>
              </div>
              <div
                aria-labelledby="gbp-list-item-8"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                <div
                  class="v-list-item__content"
                >
                  <div
                    class="v-list-item__title"
                    id="gbp-list-item-8"
                  >
                    GBP
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`CurrencyInput.vue should render correctly with value 1`] = `
<div
  class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--single-line v-select"
  eager="true"
>
  <div
    class="v-input__control"
  >
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-owns="list-16"
      class="v-input__slot"
      role="button"
    >
      <div
        class="v-select__slot"
      >
        <div
          class="v-select__selections"
        >
          <div
            class="v-select__selection v-select__selection--comma"
          >
            EUR
          </div>
          <input
            aria-readonly="true"
            id="input-16"
            readonly="readonly"
            type="text"
          />
        </div>
        <div
          class="v-input__append-inner"
        >
          <div
            class="v-input__icon v-input__icon--append"
          >
            <i
              aria-hidden="true"
              class="v-icon notranslate material-icons theme--light"
            >
              arrow_drop_down
            </i>
          </div>
        </div>
      </div>
      <div
        class="v-menu"
      >
        <div
          class="v-menu__content theme--light v-menu__content--auto "
          style="max-height: 200px; min-width: 0px; max-width: auto; top: 12px; left: 12px; transform-origin: top left; z-index: 0; display: none;"
        >
          <div
            class="v-select-list v-card theme--light"
            id="list-16"
          >
            <div
              class="v-list v-sheet v-sheet--tile theme--light"
              id="list-16"
              role="listbox"
              tabindex="-1"
            >
              <div
                aria-labelledby="eur-list-item-21"
                aria-selected="true"
                class="v-list-item primary--text v-list-item--active v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                <div
                  class="v-list-item__content"
                >
                  <div
                    class="v-list-item__title"
                    id="eur-list-item-21"
                  >
                    EUR
                  </div>
                </div>
              </div>
              <div
                aria-labelledby="usd-list-item-21"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                <div
                  class="v-list-item__content"
                >
                  <div
                    class="v-list-item__title"
                    id="usd-list-item-21"
                  >
                    USD
                  </div>
                </div>
              </div>
              <div
                aria-labelledby="gbp-list-item-21"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                <div
                  class="v-list-item__content"
                >
                  <div
                    class="v-list-item__title"
                    id="gbp-list-item-21"
                  >
                    GBP
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`CurrencyInput.vue should trigger input event if user selects new item 1`] = `
<div
  class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty v-input--is-focused theme--light v-text-field v-text-field--single-line v-text-field--is-booted v-select primary--text"
>
  <div
    class="v-input__control"
  >
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-owns="list-29"
      class="v-input__slot"
      role="button"
    >
      <div
        class="v-select__slot"
      >
        <div
          class="v-select__selections"
        >
          <div
            class="v-select__selection v-select__selection--comma"
          >
            EUR
          </div>
          <input
            aria-readonly="true"
            id="input-29"
            readonly="readonly"
            type="text"
          />
        </div>
        <div
          class="v-input__append-inner"
        >
          <div
            class="v-input__icon v-input__icon--append"
          >
            <i
              aria-hidden="true"
              class="v-icon notranslate material-icons theme--light primary--text"
            >
              arrow_drop_down
            </i>
          </div>
        </div>
      </div>
      <div
        class="v-menu"
      >
        <div
          class="v-menu__content theme--light v-menu__content--auto "
          style="max-height: 200px; min-width: 0px; max-width: auto; top: 12px; left: 12px; transform-origin: top left; z-index: 8; display: none; z-index: 8;"
        >
          <div
            class="v-select-list v-card theme--light"
            id="list-29"
          >
            <div
              class="v-list v-sheet v-sheet--tile theme--light"
              id="list-29"
              role="listbox"
              tabindex="-1"
            >
              <div
                aria-labelledby="eur-list-item-34"
                aria-selected="true"
                class="v-list-item primary--text v-list-item--active v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                <div
                  class="v-list-item__content"
                >
                  <div
                    class="v-list-item__title"
                    id="eur-list-item-34"
                  >
                    EUR
                  </div>
                </div>
              </div>
              <div
                aria-labelledby="usd-list-item-34"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                <div
                  class="v-list-item__content"
                >
                  <div
                    class="v-list-item__title"
                    id="usd-list-item-34"
                  >
                    USD
                  </div>
                </div>
              </div>
              <div
                aria-labelledby="gbp-list-item-34"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                <div
                  class="v-list-item__content"
                >
                  <div
                    class="v-list-item__title"
                    id="gbp-list-item-34"
                  >
                    GBP
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`CurrencyInput.vue should update v-model correctly 1`] = `
<div
  class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty v-input--is-focused theme--light v-text-field v-text-field--single-line v-text-field--is-booted v-select primary--text"
  eager=""
>
  <div
    class="v-input__control"
  >
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-owns="list-56"
      class="v-input__slot"
      role="button"
    >
      <div
        class="v-select__slot"
      >
        <div
          class="v-select__selections"
        >
          <div
            class="v-select__selection v-select__selection--comma"
          >
            EUR
          </div>
          <input
            aria-readonly="true"
            id="input-56"
            readonly="readonly"
            type="text"
          />
        </div>
        <div
          class="v-input__append-inner"
        >
          <div
            class="v-input__icon v-input__icon--append"
          >
            <i
              aria-hidden="true"
              class="v-icon notranslate material-icons theme--light primary--text"
            >
              arrow_drop_down
            </i>
          </div>
        </div>
      </div>
      <div
        class="v-menu"
      >
        <div
          class="v-menu__content theme--light v-menu__content--auto "
          style="max-height: 200px; min-width: 0px; max-width: auto; top: 12px; left: 12px; transform-origin: top left; z-index: 8; display: none; z-index: 8;"
        >
          <div
            class="v-select-list v-card theme--light"
            id="list-56"
          >
            <div
              class="v-list v-sheet v-sheet--tile theme--light"
              id="list-56"
              role="listbox"
              tabindex="-1"
            >
              <div
                aria-labelledby="eur-list-item-61"
                aria-selected="true"
                class="v-list-item primary--text v-list-item--active v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                <div
                  class="v-list-item__content"
                >
                  <div
                    class="v-list-item__title"
                    id="eur-list-item-61"
                  >
                    EUR
                  </div>
                </div>
              </div>
              <div
                aria-labelledby="usd-list-item-61"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                <div
                  class="v-list-item__content"
                >
                  <div
                    class="v-list-item__title"
                    id="usd-list-item-61"
                  >
                    USD
                  </div>
                </div>
              </div>
              <div
                aria-labelledby="gbp-list-item-61"
                aria-selected="false"
                class="v-list-item v-list-item--link theme--light"
                role="option"
                tabindex="0"
              >
                <div
                  class="v-list-item__content"
                >
                  <div
                    class="v-list-item__title"
                    id="gbp-list-item-61"
                  >
                    GBP
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
